File: backend/app/schemas/updates.py
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field


class UpdateStatusPublic(BaseModel):
    current: str
    available: str | None
    target_version: str | None
    channel: str
    repo: str
    state: Literal[
        "idle",
        "checking",
        "backing_up",
        "pulling",
        "recreating",
        "health_checking",
        "success",
        "failed",
    ]
    last_checked_iso: str | None
    last_error: str | None
    # Set when the release poll failed. Distinguishes "you are on the newest
    # release" from "we could not find out" — `available: null` means both.
    # Defaulted so an older updater container, which does not send the
    # field, still validates during the window between image recreations.
    check_error: str | None = None
    started_at_iso: str | None
    finished_at_iso: str | None
    log_tail: list[str]


class TriggerUpdateRequest(BaseModel):
    version: str = Field(min_length=1, max_length=64)
    backup: bool = True

File: frontend/src/lib/updates.ts
import { api, readErrorMessage } from "@/lib/api";

export type UpdateState =
  | "idle"
  | "checking"
  | "backing_up"
  | "pulling"
  | "recreating"
  | "health_checking"
  | "success"
  | "failed";

export interface UpdateStatus {
  current: string;
  available: string | null;
  target_version: string | null;
  channel: string;
  repo: string;
  state: UpdateState;
  last_checked_iso: string | null;
  last_error: string | null;
  started_at_iso: string | null;
  finished_at_iso: string | null;
  log_tail: string[];
}

export const IN_PROGRESS_STATES: UpdateState[] = [
  "backing_up",
  "pulling",
  "recreating",
  "health_checking",
];

export function isInProgress(s: UpdateState): boolean {
  return IN_PROGRESS_STATES.includes(s);
}

export async function getUpdateStatus(force = false): Promise<UpdateStatus> {
  return api
    .get("system/update/status", {
      searchParams: force ? { force: "true" } : undefined,
      // Force checks go through the list-mode path on GitHub and so can
      // take a few seconds; widen the timeout vs the default poll.
      timeout: force ? 30_000 : 15_000,
    })
    .json<UpdateStatus>();
}

export async function triggerUpdate(
  version: string,
  backup = true,
): Promise<{ status: string; target_version: string }> {
  try {
    return await api
      .post("system/update", { json: { version, backup }, timeout: 30_000 })
      .json<{ status: string; target_version: string }>();
  } catch (e) {
    throw new Error(await readErrorMessage(e));
  }
}
